src/main/java/mekanism/common/base/target/Target.java

```diff
@@ -0,0 +1,109 @@
+ * @param <TYPE> The type that is being transferred.
+ * @param <EXTRA> Any extra information this target may need to keep track of.
+ */
+public abstract class Target<HANDLER, TYPE extends Number & Comparable<TYPE>, EXTRA> {
+
+    /**
+     * Map of the sides to the handler for that side.
+     */
+    protected final Map<EnumFacing, HANDLER> handlers = new EnumMap<>(EnumFacing.class);
+    /**
+     * Map of sides that want more than we can/are willing to provide. Value is the amount they want.
+     */
+    protected final Map<EnumFacing, TYPE> needed = new EnumMap<>(EnumFacing.class);
+
+    protected EXTRA extra;
+
+    public void addHandler(EnumFacing side, HANDLER handler) {
+        handlers.put(side, handler);
+    }
+
+    public Map<EnumFacing, HANDLER> getHandlers() {
+        return handlers;
+    }
+
+    /**
+     * Sends the remaining amount to each handler we still have not settled on an amount for. We increment the amount
+     * sent in splitInfo as well as adjust the split as needed if one ends up accepting less than it originally wanted.
+     * (The most likely case this would
+     *
+     * @param splitInfo Keeps track of the current amount sent and the default each one can get.
+     */
+    public void sendRemainingSplit(SplitInfo<TYPE> splitInfo) {
+        //If needed is not empty then we default it to the given calculated fair split amount of remaining energy
+        for (EnumFacing side : needed.keySet()) {
+            acceptAmount(side, splitInfo, splitInfo.getAmountPer());
+        }
+    }
+
+    /**
+     * Gives the handler on the specified side the given amount.
+     *
+     * @param side Side of handler to give.
+     * @param splitInfo Information about current overall split. The given split will be increased by the actual amount
+     * accepted, in case it is less than the offered amount.
+     * @param amount Amount to give.
+     */
+    protected abstract void acceptAmount(EnumFacing side, SplitInfo<TYPE> splitInfo, TYPE amount);
+
+    /**
+     * Simulate inserting into the handler.
+     *
+     * @param handler The handler (should correspond with the side we are simulating).
+     * @param side The side we are simulating
+     * @param extra All the information we are inserting.
+     * @return The amount it was actually willing to accept.
+     */
+    protected abstract TYPE simulate(HANDLER handler, EnumFacing side, EXTRA extra);
+
+    public void simulate(EXTRA toSend, SplitInfo<TYPE> splitInfo) {
+        for (Entry<EnumFacing, HANDLER> entry : handlers.entrySet()) {
+            TYPE amountNeeded = simulate(entry.getValue(), entry.getKey(), toSend);
+            if (amountNeeded.compareTo(splitInfo.getAmountPer()) <= 0) {
+                //Add the amount, in case something changed from simulation only mark actual sent amount
+                // in split info
+                acceptAmount(entry.getKey(), splitInfo, amountNeeded);
```

Comment:
inside a `simulate` method, but the `acceptAmount` implementations are for actually sending?

Reply:
Renamed the method and added JavaDocs of how it works.